aiosmtpd/tests/test_smtp.py

```diff
@@ -1,86 +1,101 @@
 """Test the SMTP protocol."""
 
+import os
 import time
+import pytest
 import socket
 import asyncio
-import unittest
+import logging
+import itertools
 
 from aiosmtpd.controller import Controller
 from aiosmtpd.handlers import Sink
-from aiosmtpd.smtp import MISSING, SMTP as Server, __ident__ as GREETING
+from aiosmtpd.smtp import (
+    MISSING,
+    Session as SMTPSession,
+    SMTP as Server,
+    __ident__ as GREETING,
+)
 from aiosmtpd.testing.helpers import (
+    ReceivingHandler,
     SUPPORTED_COMMANDS_NOTLS,
-    assert_auth_invalid,
-    assert_auth_required,
-    assert_auth_success,
     reset_connection,
 )
+from aiosmtpd.testing.statuscodes import SMTP_STATUS_CODES as S
+from .conftest import Global
 from base64 import b64encode
-from contextlib import ExitStack
-from smtplib import (
-    SMTP, SMTPDataError, SMTPResponseException, SMTPServerDisconnected)
-from unittest.mock import Mock, PropertyMock, patch
+from contextlib import suppress
+from smtplib import SMTP, SMTPDataError, SMTPResponseException, SMTPServerDisconnected
+from textwrap import dedent
+from typing import AnyStr, Callable, List, Tuple
+from unittest.mock import MagicMock
+
+
+CRLF = "\r\n"
+BCRLF = b"\r\n"
+MAIL_LOG = logging.getLogger("mail.log")
 
-CRLF = '\r\n'
-BCRLF = b'\r\n'
+ASYNCIO_CATCHUP_DELAY = float(os.environ.get("ASYNCIO_CATCHUP_DELAY", 0.1))
+"""
+Delay (in seconds) to give asyncio event loop time to catch up and do things. May need
+to be increased for slow and/or overburdened test systems.
+"""
```

Comment:
I found out that pytest works too fast not allowing asyncio to catch up during some tests. This delay yields the main thread to asyncio so the event loop can catch up and behave as expected.